core/templates/dev/head/pages/exploration-player-page/services/learner-answer-info.service.ts

```diff
@@ -0,0 +1,116 @@
+// Copyright 2019 The Oppia Authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS-IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/**
+ * @fileoverview Service for learner answer info.
+ */
+
+require(
+  'pages/exploration-player-page/services/answer-classification.service.ts');
+require('domain/statistics/LearnerAnswerDetailsBackendApiService.ts');
+
+var oppia = require('AppInit.ts').module;
+
+oppia.factory('LearnerAnswerInfoService', [
+  'AnswerClassificationService', 'LearnerAnswerDetailsBackendApiService',
+  'INTERACTION_IDS_WITHOUT_ANSWER_DETAILS', 'PROBABILITY_INDEXES',
+  function(
+      AnswerClassificationService, LearnerAnswerDetailsBackendApiService,
+      INTERACTION_IDS_WITHOUT_ANSWER_DETAILS, PROBABILITY_INDEXES) {
+    var submittedAnswerInfoCount = 0;
+    var currentEntityId = null;
+    var stateName = null;
+    var interactionId = null;
+    var currentAnswer = null;
+    var currentInteractionRulesService = null;
+    var canAskLearnerForAnswerInfo = false;
+    var visitedStates = [];
+
+
+    var getRandomProbabilityIndex = function() {
+      var min = 0;
+      var max = 100;
+      return (Math.floor(Math.random() * (max - min + 1) ) + min) / 100;
+    };
+
+    return {
+      evalAskLearnerForAnswerInfo: function(
```

Comment:
Not sure I understand the difference between this and canAskLearnerForAnswerInfo(), and when the two methods are expected to be called.

Can you give a description of the "lifecycle" of calls to this service? I.e. explain, from the point where a learner starts playing an exploration, what methods are supposed to get called. How often do we need to do this evaluation from scratch? Also, when is the cached canAskLearnerForAnswerInfo variable supposed to be used -- is it only for the current state or is the idea that it persists through multiple states? 

(This will help me suggest better structure/naming.)

Reply:
Actually, when the learner submits the answer, the LearnerAnswerInfo service is initiated i.e. it evaluates according to the answer submitted by the learner that whether the learner should be asked for the answer details or not i.e. it marks the canAskLearnerForAnswerInfo true if the result of the evaluation is true or false its the evaluation is false. The evaluation result depends on the outcome of the answer. After the initiation, the canAskLearnerForAnswerInfo is checked if it's true the learner is asked for answer details otherwise not.

As per now, the learner will be asked for the answer details for two times per exploration.